packages/ember-application/lib/system/application.js

```diff
@@ -318,24 +364,25 @@ var Application = Namespace.extend(RegistryProxy, {
     registerLibraries();
     logLibraryVersions();
 
-    // Start off the number of deferrals at 1. This will be
-    // decremented by the Application's own `initialize` method.
+    // Start off the number of deferrals at 1. This will be decremented by
+    // the Application's own `boot` method.
     this._readinessDeferrals = 1;
 
     if (isEnabled('ember-application-visit')) {
-      if (this.autoboot) {
-        // Create subclass of Ember.Router for this Application instance.
-        // This is to ensure that someone reopening `App.Router` does not
-        // tamper with the default `Ember.Router`.
-        // 2.0TODO: Can we move this into a globals-mode-only library?
-        this.Router = (this.Router || Router).extend();
-        this.buildDefaultInstance();
-        this.waitForDOMReady();
-      }
+      this.autoboot = this._globalsMode = !!this.autoboot;
     } else {
-      this.Router = (this.Router || Router).extend();
-      this.buildDefaultInstance();
-      this.waitForDOMReady();
+      // Force-assign these flags to their default values when the feature is
+      // disabled, this ensures we can rely on their values in other paths.
+      this.autoboot = true;
+      this._globalsMode = true;
+    }
+
+    if (this._globalsMode) {
+      this._prepareForGlobalsMode();
+    }
+
+    if (this.autoboot) {
+      this._scheduleAutoBoot();
```

Comment:
@kratiahuja I'm not sure if needing to boot the app on `init()` is common-enough of a use case that we need to privilege it with more public API.